lib/profile.dart

```diff
@@ -1,23 +1,145 @@
 import 'package:flutter/material.dart';
 
+import 'appstate.dart';
+import 'strings.dart';
+
+// ignore: must_be_immutable
 class UserProfilePage extends StatefulWidget {
-  final String title = 'Profile';
+  AppState appState;
 
-  const UserProfilePage({Key? key}) : super(key: key);
+  UserProfilePage({Key? key, required this.appState}) : super(key: key);
 
   @override
   State<UserProfilePage> createState() => _UserProfilePageState();
 }
 
 class _UserProfilePageState extends State<UserProfilePage> {
+  final _formKey = GlobalKey<FormState>();
+  final _statusMessageInputController = TextEditingController();
+  final _nickInputController = TextEditingController();
+  bool _applyButtonPressed = false;
+
+  void _onUpdateNick() {
+    if (_formKey.currentState!.validate()) {
+      widget.appState.setNick(_nickInputController.text);
+    }
+  }
+
+  void _onUpdateStatusMessage() {
+    if (_formKey.currentState!.validate()) {
+      widget.appState.setUserStatus(_statusMessageInputController.text);
+    }
+  }
+
+  bool _nickIsValid(String nick) {
+    return _nickInputController.text.isNotEmpty &&
+        _nickInputController.text.length <= 32;
+  }
+
+  bool _statusMessageIsValid(String nick) {
+    return _statusMessageInputController.text.length <= 256;
+  }
+
+  _onValidate() {
+    if (!_nickIsValid(_nickInputController.text) ||
```

Comment:
You can use `_formKey.currentState.validate()` to run every validation function registered in the form. That way you don't need any helpers for validation anymore.

Reply:
The onValidate method is what pushes the button and sets the state. We don't want to call that every keystroke. Unless I'm missing something...